core/services/keeper/registry_synchronizer_test.go

```diff
@@ -0,0 +1,130 @@
+	j := cltest.MustInsertKeeperJob(t, store, cltest.NewEIP55Address(), cltest.NewEIP55Address())
+	contractAddress := j.KeeperSpec.ContractAddress
+	contract, err := keeper_registry_wrapper.NewKeeperRegistry(
+		contractAddress.Address(),
+		ethMock,
+	)
+	require.NoError(t, err)
+
+	synchronizer := keeper.NewRegistrySynchronizer(j, contract, keeperORM, syncInterval)
+	return store, synchronizer, ethMock, j, cleanup
+}
+
+func Test_RegistrySynchronizer_Start(t *testing.T) {
+	_, synchronizer, _, _, cleanup := setupRegistrySync(t)
+	defer cleanup()
+
+	err := synchronizer.Start()
+	require.NoError(t, err)
+	defer synchronizer.Close()
+
+	err = synchronizer.Start()
+	require.Error(t, err)
+}
+
+func Test_RegistrySynchronizer_AddsAndRemovesUpkeeps(t *testing.T) {
+	store, synchronizer, ethMock, job, cleanup := setupRegistrySync(t)
+	defer cleanup()
+
+	contractAddress := job.KeeperSpec.ContractAddress.Address()
+	fromAddress := job.KeeperSpec.FromAddress.Address()
+
+	// 1st sync
+	registryMock := cltest.NewContractMockReceiver(t, ethMock, keeper.RegistryABI, contractAddress)
+	canceledUpkeeps := []*big.Int{big.NewInt(1)}
+	registryMock.MockResponse("getConfig", regConfig).Once()
+	registryMock.MockResponse("getKeeperList", []common.Address{fromAddress}).Once()
+	registryMock.MockResponse("getCanceledUpkeepList", canceledUpkeeps).Once()
+	registryMock.MockResponse("getUpkeepCount", big.NewInt(3)).Once()
+	registryMock.MockResponse("getUpkeep", upkeep).Times(3) // sync all 3, then delete
+
+	synchronizer.ExportedSyncRegistry()
+
+	cltest.AssertCount(t, store, keeper.Registry{}, 1)
+	cltest.AssertCount(t, store, keeper.UpkeepRegistration{}, 2)
+
+	var registry keeper.Registry
+	var upkeepRegistration keeper.UpkeepRegistration
+	require.NoError(t, store.DB.First(&registry).Error)
+	require.NoError(t, store.DB.First(&upkeepRegistration).Error)
+	require.Equal(t, job.KeeperSpec.ContractAddress, registry.ContractAddress)
+	require.Equal(t, job.KeeperSpec.FromAddress, registry.FromAddress)
+	require.Equal(t, int32(20), registry.BlockCountPerTurn)
+	require.Equal(t, int32(0), registry.KeeperIndex)
+	require.Equal(t, int32(1), registry.NumKeepers)
+	require.Equal(t, upkeep.CheckData, upkeepRegistration.CheckData)
+	require.Equal(t, int32(upkeep.ExecuteGas), upkeepRegistration.ExecuteGas)
+
+	ethMock.AssertExpectations(t)
+
+	// 2nd sync
+	canceledUpkeeps = []*big.Int{big.NewInt(0), big.NewInt(1), big.NewInt(3)}
+	registryMock.MockResponse("getConfig", regConfig).Once()
+	registryMock.MockResponse("getKeeperList", []common.Address{fromAddress}).Once()
+	registryMock.MockResponse("getCanceledUpkeepList", canceledUpkeeps).Once()
+	registryMock.MockResponse("getUpkeepCount", big.NewInt(5)).Once()
+	registryMock.MockResponse("getUpkeep", upkeep).Times(2) // two new upkeeps to sync
+
+	synchronizer.ExportedSyncRegistry()
+
+	cltest.AssertCount(t, store, keeper.Registry{}, 1)
+	cltest.AssertCount(t, store, keeper.UpkeepRegistration{}, 2)
```

Comment:
Can we sanity check the first and second DB records here? (just to make sure it didn't get overwritten or something daft)